### PR TITLE
Fix for Brains not reinitialising when the scene is reloaded.

### PR DESCRIPTION
**Pull request by tkggwatson**
This was a bug caused by the conversion of Brains over to ScriptableObjects. ScriptableObjects persist in memory between scene changes, which means that after a scene change the Brains would still be initialised and the agentInfos list would contain invalid references to the Agents from the previous scene.

The fix is to have the Academy notify the Brains when it is destroyed. This allows the Brains to clean themselves up and transition back to an uninitialised state. After the new scene is loaded, the Brain's LazyInitialise will reconnect the Brain to the new Academy as expected.

### CODE REVIEW

**Comment by vincentpierre** [comment]
Hi @tkggwatson, thank you for this PR.
The Academy was not meant to be destroyed and reinitialized between scene loads. Would calling 

```csharp
DontDestroyOnLoad(academyGameObject);
```
Before loading the next scene solve your problem ?

**Comment by tkggwatson** [comment]
Hi Vincent,

The academy is currently a scene object, and it is entirely feasible to have a different academy for different scenes (e.g. a game with different level types that need different brains and derived academy parameters).

Adding a DontDestroyOnLoad (or just making the academy a ScriptableObject) would not entirely fix the issue, because the agents would remain registered with the brain. That could be solved by making the agents deregister themselves from the brain in OnDestroy or OnDisable, but it would still not solve the varying level design problem.

An alternative might be to get rid of academy completely... Now that Brains are ScriptableObjects they could entirely absorb the functionality of the Academy and make usage simpler overall. However I'm not in a position to contribute much to the architecture, so I'm interested in stabilizing the current implementation.

Regards
Graham

**Comment by vincentpierre** [comment]
Hi again, do you have a minimal sample project we could look at that causes the problem you are describing ? The agentInfos of the brain should have been cleared right after the decision has been make. This means that the agents requested a decision and the new scene was loaded before the agents received their actions. I would like to know in what situations this happens.

**Comment by tkggwatson** [comment]
Hi, my project is a full game, however I believe that you can just modify any of the samples to have a reload button that just loads the scene again in order to repro.

In previous versions the agent list was cleared, but now the Brain is a ScriptableObject the in-memory copy persists between scene loads, so the brain stays initialised whilst the new scene is loading.

**Comment by tkggwatson** [comment]
See bug https://github.com/Unity-Technologies/ml-agents/issues/1754 which indicates that others are also finding this issue.

File: UnitySDK/Assets/ML-Agents/Scripts/Academy.cs [hunk 1]
```diff
@@ -230,6 +233,12 @@ void Awake()
             InitializeEnvironment();
         }
 
+        private void OnDestroy()
```
Comment (vincentpierre) [inline UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: In the develop branch, the method OnDestroy is used to reset some physics parameters, can you move DestroyAction() to this new method ?

**Comment by tkggwatson** [comment]
Oops, think I did something wrong when trying to rebase this to develop. Am I supposed to reopen the pull request @vincentpierre ? Apologies but Git isn't my main source control system.

**Comment by tkggwatson** [comment]
Okay @vincentpierre I think I did it right! Your suggested changes have been made and I've rebased it to develop.

File: UnitySDK/Assets/ML-Agents/Scripts/Academy.cs (hunk 1 above)
Comment (tkggwatson) [reply UnitySDK/Assets/ML-Agents/Scripts/Academy.cs]: These changes have been made.

**Pull request merged**